etag/middleware.go

```diff
@@ -14,31 +15,53 @@ const (
 	IfNoneMatchHeaderKey = "If-None-Match"
 )
 
+var (
+	errDoubleWrite = errors.New("Write called twice with ETag Middleware")
+)
+
 type etagResponseWriter struct {
 	http.ResponseWriter
-	req  *http.Request
-	code int
+	req         *http.Request
+	code        int
+	wroteHeader bool
+	wroteBody   bool
 }
 
 func (erw *etagResponseWriter) Write(b []byte) (int, error) {
-	etag := etag(b)
+	etag := calculateEtag(b)
 	erw.Header().Set(ETagHeaderKey, etag)
+
 	if erw.req.Header.Get(IfNoneMatchHeaderKey) == etag {
-		erw.WriteHeader(http.StatusNotModified)
-		return erw.Write(nil)
+		erw.code = http.StatusNotModified
+		erw.writeHeader()
+		return erw.writeBody(nil)
 	}
 
-	erw.WriteHeader(erw.code)
-	return erw.ResponseWriter.Write(b)
+	erw.writeHeader()
+	return erw.writeBody(b)
 }
 
-func etag(data []byte) string {
+func calculateEtag(data []byte) string {
 	return fmt.Sprintf("%x", md5.Sum(data))
 }
 
+func (erw *etagResponseWriter) writeBody(data []byte) (int, error) {
+	if erw.wroteBody {
+		panic(errDoubleWrite)
```

Comment:
To be 100% correct, this write should be protected by a lock, or else double writes could technically still happen.